File: dotnet/base/Mcma.Core/Model/McmaResource.cs
using System;

namespace Mcma.Core
{
    public abstract class McmaResource : McmaObject
    {
        public string Id { get; set; }

        public DateTime? DateCreated { get; set; }

        public DateTime? DateModified { get; set; }

        public void OnCreate(string id)
        {
            Id = id;
            DateModified = DateCreated = DateTime.UtcNow;
        }

        public void OnUpsert(string id)
        {
            Id = id;
            DateModified = DateTime.UtcNow;
            if (!DateCreated.HasValue)
                DateCreated = DateModified;
        }
    }
}

File: nodejs/base/core/lib/model/mcma-resource.ts
import { McmaObject, McmaObjectProperties } from "./mcma-object";

export interface McmaResourceProperties extends McmaObjectProperties {
    id?: string;
    dateCreated?: Date | string;
    dateModified?: Date | string;
}

export function onResourceCreate(resource: McmaResourceProperties, id: string) {
    resource.id = id;
    resource.dateModified = resource.dateCreated = new Date();
}

export function onResourceUpsert(resource: McmaResourceProperties, id: string) {
    resource.id = id;
    resource.dateModified = new Date();
    if (!resource.dateCreated) {
        resource.dateCreated = resource.dateModified;
    }
}

export abstract class McmaResource extends McmaObject implements McmaResourceProperties {
    id?: string;
    dateCreated?: Date;
    dateModified?: Date;

    protected constructor(type: string, properties: McmaResourceProperties) {
        super(type, properties);

        this.dateCreated = this.ensureValidDateOrUndefined(this.dateCreated);
        this.dateModified = this.ensureValidDateOrUndefined(this.dateModified);
    }

    onCreate = (id: string) => onResourceCreate(this, id);
    onUpsert = (id: string) => onResourceUpsert(this, id);
}
